crates/client/rpc/src/lib.rs

```diff
@@ -477,10 +483,81 @@ where
         };
         Ok(MaybePendingBlockWithTxHashes::Block(block_with_tx_hashes))
     }
-}
 
-/// Removes the "0x" prefix from a given hexadecimal string and pads it with 0s
-#[inline(always)]
-fn format_hex(input: &str) -> String {
-    format!("{:0>64}", input.strip_prefix("0x").unwrap_or(input))
+    /// Declares a new contract class
+    ///
+    /// # Arguments
+    ///
+    /// * `declare_transaction` - The declare transaction
+    ///
+    /// # Returns
+    ///
+    /// * `transaction_hash` - The hash of the transaction
+    /// * `class_hash` - The hash of the declared class
+    fn add_declare_transaction(
+        &self,
+        declare_transaction: Transaction,
+        block_id: StarknetBlockId,
+    ) -> RpcResult<AddDeclareTransactionOutput> {
+        let tx: BroadcastedDeclareTransactionV2 = match declare_transaction {
+            Transaction::Declare(declare) => Ok(declare),
+            _ => {
+                error!("Transaction type should be DECLARE.");
+                Err(StarknetRpcApiError::InternalServerError)
+            }
+        }?;
+
+        let substrate_block_hash = self.substrate_block_hash_from_starknet_block(block_id).map_err(|e| {
+            error!("'{e}'");
+            StarknetRpcApiError::BlockNotFound
+        })?;
+
+        let runtime_api = self.client.runtime_api();
+
+        // Program is send as gzip + base64 encoded, we need to decompress it
+        // Decode the base64 encoded string
+        let compressed_bytes = base64::decode(&tx.contract_class.program).unwrap();
+
+        // Create a GzipDecoder to decompress the bytes
+        let mut gz = GzDecoder::new(&compressed_bytes[..]);
+
+        // Read the decompressed bytes into a Vec<u8>
+        let mut decompressed_bytes = Vec::new();
+        gz.read_to_end(&mut decompressed_bytes).unwrap();
+
+        // Deserialize the decompressed bytes into a Rust struct using serde_json
+        let program: ProgramWrapper = serde_json::from_slice(&decompressed_bytes).unwrap();
+
+        // TODO: Add support for V2 Declare TXs
+        // TODO: remove unwraps
```

Comment:
since this is a draft PR it's fine for now, but just want to make sure we don't forget it when you change the status of the PR, so I am adding a comment